src/v/storage/segment_appender.cc

```diff
@@ -445,66 +446,113 @@ void segment_appender::dispatch_background_head_write() {
+             */
+            _prev_head_write = ss::make_lw_shared<ssx::semaphore>(
+              1, head_sem_name);
+        }
+    });
+
+    if (!_inflight.empty() && _inflight.back()->try_merge(entry, prior_co)) {
+        // Yay! The latest in-flight write is still queued (i.e., has not
+        // been dispatched to the disk) so we just append this write
+        // to that entry.
+        return;
+    }
+
     _inflight.emplace_back(
-      ss::make_lw_shared<inflight_write>(_committed_offset));
-    auto w = _inflight.back();
+      ss::make_lw_shared<inflight_write>(std::move(entry)));
 
-    /*
-     * if _head is full then take control of it for this final write and then
-     * release it back into the chunk cache. otherwise, leave it in place so
-     * that new appends may accumulate. this optimization is meant to avoid
-     * redhydrating the chunk on append following a flush when the head has
-     * pending bytes and a write is dispatched.
-     */
-    const auto full = _head->is_full();
-    auto h = full ? std::exchange(_head, nullptr) : _head;
+    auto w = _inflight.back();
 
     /*
-     * make sure that when the write is dispatched that is sequenced in-order on
-     * the correct semaphore by grabbing the units synchronously.
+     * make sure that when the write is dispatched that is sequenced
+     * in-order on the correct semaphore by grabbing the units
+     * synchronously.
      */
     auto prev = _prev_head_write;
     auto units = ss::get_units(*prev, 1);
 
     (void)ss::with_semaphore(
       _concurrent_flushes,
       1,
-      [h,
-       w,
-       this,
-       start_offset,
-       expected,
-       src,
-       prev,
-       units = std::move(units),
-       full]() mutable {
+      [w, this, prev, units = std::move(units)]() mutable {
           return units
-            .then([this, h, w, start_offset, expected, src, full](
-                    ssx::semaphore_units u) mutable {
+            .then([this, w](ssx::semaphore_units u) mutable {
+                const auto dma_size = w->chunk_end - w->chunk_begin;
+
+                vassert(
+                  dma_size <= _chunk_size && w->chunk_end > w->chunk_begin
+                    && w->chunk_end <= _chunk_size,
+                  "Bad write bounds _chunk_size: {}, chunk_begin: {}, "
+                  "chunk_end: {}",
+                  _chunk_size,
+                  w->chunk_begin,
+                  w->chunk_end);
+
+                // prevent any more writes from merging into this entry
+                // as it is about to be dma_write'd.
+                w->set_state(DISPATCHED);
+
                 return _out
-                  .dma_write(start_offset, src, expected, _opts.priority)
-                  .then([this, h, w, expected, full](size_t got) {
+                  .dma_write(
+                    w->file_start_offset,
+                    w->chunk->data() + w->chunk_begin,
+                    w->chunk_end - w->chunk_begin,
```

Comment:
The args to `dma_write` are the main reason we need to move stuff into the `w` struct as the size of the write may change. So we read the correct size from `w` at the moment we dispatch.